hazelcast/src/test/java/com/hazelcast/client/connectionstrategy/ConfiguredBehaviourTest.java

```diff
@@ -113,6 +116,40 @@ public void testAsyncStartTrue() {
         });
     }
 
+    @Test
+    public void testAsyncStartTrueShouldNotBlock_whenThereIsClientStateToSend() {
+        Config config = new Config();
+        config.getUserCodeDeploymentConfig().setEnabled(true);
+        hazelcastFactory.newHazelcastInstance(config);
+
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.getConnectionStrategyConfig().getConnectionRetryConfig().setClusterConnectTimeoutMillis(-1);
+        ClientUserCodeDeploymentConfig clientUserCodeDeploymentConfig = new ClientUserCodeDeploymentConfig();
+        clientUserCodeDeploymentConfig.addClass(IncrementingEntryProcessor.class);
+        clientUserCodeDeploymentConfig.setEnabled(true);
+        clientConfig.setUserCodeDeploymentConfig(clientUserCodeDeploymentConfig);
+        clientConfig.getConnectionStrategyConfig().setAsyncStart(true);
+        HazelcastInstance client = hazelcastFactory.newHazelcastClient(clientConfig);
+
+        // The following logic verifies user code deployment works
```

Comment:
You need to use `FilteringClassLoader` on the member config to make sure that it is using the class that is sent with UCD. Otherwise, since the member and client are in the same JVM, it will work anyways. See `ClientUserCodeDeploymentTest#createNodeConfig`